File: 202407/2196.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        d = {}
        children = set()
        for p, c, l in descriptions:
            if p not in d:
                d[p] = TreeNode(p)
            if c not in d:
                d[c] = TreeNode(c)
            children.add(c)
            if l == 1:
                d[p].left = d[c]
            else:
                d[p].right = d[c]
        
        for p, _, _ in descriptions:
            if p not in children:
                return d[p]
        return None

File: 202606/2196.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, TreeNode*> nodes;
        for(const auto& des : descriptions) {
            TreeNode* child = nullptr;
            if (nodes.find(des[1]) == nodes.end()) {
                child = new TreeNode(des[1]);
                nodes[des[1]] = child;
            } else {
                child = nodes[des[1]];
            }
            TreeNode* parent = nullptr;
            if (nodes.find(des[0]) == nodes.end()) {
                parent = new TreeNode(des[0]);
                nodes[des[0]] = parent;
            } else {
                parent = nodes[des[0]];
            }
            if (des[2] == 0) {
                parent->right = child;
            } else {
                parent->left = child;
            }
        }
        for(const auto& des : descriptions) {
            nodes.erase(des[1]);
        }
        return nodes.begin()->second;
    }
};
